Formal2024/Workshops/workshop3.lean

```diff
@@ -0,0 +1,210 @@
+/-
+  # Logic: Conjunction, Iff, and Disjunction
+  References: [MIL] Mathematics in Lean, [Tut] Tutorials project.
+  Most of the demonstration section comes [MIL].
+-/
+
+import Mathlib.Tactic
+
+/-!
+  # Conjunction
+-/
+
+-- The symbol for conjunction is `∧`, which you can write as `\and`.
+
+-- To prove a conjunction, we use the `constructor` tactic.
+-- This will split the goal into two, one for each part of the conjunction.
+-- (For people familiar with Lean 3: the `split` tactic is gone in Lean 4.)
+
+example (P Q : Prop) (hp : P) (hq : Q) : P ∧ Q := by
+  constructor
+  · exact hp
+  · exact hq
+
+-- Here is another example:
+example {x y : ℝ} (h₀ : x ≤ y) (h₁ : ¬y ≤ x) : x ≤ y ∧ x ≠ y := by
+  constructor
+  · assumption -- This tactic looks for a hypothesis in the context that matches the goal, like `h₀`.
+  · intro h
+    apply h₁
+    rw [h]
+
+-- You can also use the anonymous constructor `⟨⟩`:
+example {x y : ℝ} (h₀ : x ≤ y) (h₁ : ¬y ≤ x) : x ≤ y ∧ x ≠ y := by
+  have h : x ≠ y := by
+    contrapose! h₁
+    rw [h₁]
+  exact ⟨h₀, h⟩
+
+
+-- To use a conjunction, you can use `rcases` or `rintro`, as with existential quantifiers.
+example {x y : ℝ} (h : x ≤ y ∧ x ≠ y) : ¬y ≤ x := by
+  rcases h with ⟨h₀, h₁⟩
+  contrapose! h₁
+  exact le_antisymm h₀ h₁
+
+example {x y : ℝ} : x ≤ y ∧ x ≠ y → ¬y ≤ x := by
```

Comment:
Just flagging that this example will be hard for beginner learners to parse! They will need to figure out
a) why rintro ends up producing a goal of `False`
b) why the `exact` tactic make sense applied to these arguments.

Because they will be new to working with negation, these things will not be obvious. I think it's a good example to illustrate these concepts, just wanted to flag that you may want to break it down quite slowly in class with them.